@mnshdw Regarding why we create a new tailer: there was an issue with a double close of a channel. I think @ajacquemot already had this issue in the udp tailer. We figured it was easier to create a new tailer and let the older get GCed 

```
2018-09-13 12:57:04 UTC | INFO | (tailer.go:93 in Start) | Start tailing container: da1e7badfbe8
2018-09-13 12:57:05 UTC | WARN | (launcher.go:219 in restartTailer) | Could not start tailer for container da1e7badfbe8: error during connect: Get http://%2Fvar%2Frun%2Fdocker.sock/v1.25/containers/da1e7badfbe81904fbf40de8aa2629aaf648183e054e6db0c1d1e75e11c3e123/logs?follow=1&since=1536843424.788822639&stderr=1&stdout=1&tail=&timestamps=1: read unix @->/var/run/docker.sock: read: connection reset by peer
2018-09-13 12:57:06 UTC | INFO | (tailer.go:93 in Start) | Start tailing container: da1e7badfbe8
panic: close of closed channel

goroutine 228 [running]:
github.com/DataDog/datadog-agent/pkg/logs/decoder.(*Decoder).Stop(0xc4209eeed0)
	/home/vagrant/go/src/github.com/DataDog/datadog-agent/pkg/logs/decoder/decoder.go:96 +0x2e
github.com/DataDog/datadog-agent/pkg/logs/input/docker.(*Tailer).readForever(0xc420172b80)
	/home/vagrant/go/src/github.com/DataDog/datadog-agent/pkg/logs/input/docker/tailer.go:141 +0x4fb
created by github.com/DataDog/datadog-agent/pkg/logs/input/docker.(*Tailer).tail
	/home/vagrant/go/src/github.com/DataDog/datadog-agent/pkg/logs/input/docker/tailer.go:127 +0x13f
```

--- a/pkg/logs/input/docker/launcher.go
+++ b/pkg/logs/input/docker/launcher.go
@@ -8,6 +8,8 @@
 package docker
 
 import (
+	"time"
+
 	"github.com/DataDog/datadog-agent/pkg/tagger"
 	"github.com/DataDog/datadog-agent/pkg/util/log"
 	"github.com/docker/docker/client"
@@ -177,24 +179,43 @@ func (l *Launcher) stopTailer(containerID string) {
 }
 
 func (l *Launcher) restartTailer(containerID string) {
-	oldTailer, _ := l.tailers[containerID]
-	source := oldTailer.source
-	oldTailer.Stop()
+	backoffDuration := 1 * time.Microsecond
+	backoffMax := 60 * time.Microsecond
+	var tailer *Tailer
+	var source *config.LogSource
 
-	tailer := NewTailer(l.cli, containerID, source, l.pipelineProvider.NextPipelineChan(), l.lostSocket)
+	for i := 0; ; i++ {
+		if backoffDuration > backoffMax {
+			backoffDuration = backoffMax
+		}
 
-	// compute the offset to prevent from missing or duplicating logs
-	since, err := Since(l.registry, tailer.Identifier(), service.Before)
-	if err != nil {
-		log.Warnf("Could not recover tailing from last committed offset: %v", ShortContainerID(containerID), err)
+		time.Sleep(backoffDuration)
+
+		if i == 0 {
+			oldTailer, _ := l.tailers[containerID]
+			source = oldTailer.source
+			oldTailer.Stop()
+		}
+
+		tailer = NewTailer(l.cli, containerID, source, l.pipelineProvider.NextPipelineChan(), l.lostSocket)
+
+		// compute the offset to prevent from missing or duplicating logs
+		since, err := Since(l.registry, tailer.Identifier(), service.Before)
+		if err != nil {
+			log.Warnf("Could not recover tailing from last committed offset: %v", ShortContainerID(containerID), err)
+			backoffDuration *= 2
+			continue
+		}
+
+		// start the tailer
+		err = tailer.Start(since)
+		if err != nil {
+			log.Warnf("Could not start tailer: %v", containerID, err)
+			backoffDuration *= 2
+			continue
+		}
+		// keep the tailer in track to stop it later on
+		l.tailers[containerID] = tailer
+		return
 	}
-
-	// start the tailer
-	err = tailer.Start(since)
-	if err != nil {
-		log.Warnf("Could not start tailer: %v", containerID, err)
-	}
-
-	// keep the tailer in track to stop it later on
-	l.tailers[containerID] = tailer
 }

--- a/pkg/logs/input/docker/launcher.go
+++ b/pkg/logs/input/docker/launcher.go
@@ -31,6 +31,7 @@ type Launcher struct {
 	cli               *client.Client
 	registry          auditor.Registry
 	stop              chan struct{}
+	lostSocket        chan string
 }
 
 // NewLauncher returns a new launcher
@@ -44,6 +45,7 @@ func NewLauncher(sources *config.LogSources, services *service.Services, pipelin
 		pendingContainers: make(map[string]*Container),
 		registry:          registry,
 		stop:              make(chan struct{}),
+		lostSocket:        make(chan string, 1),
 	}
 	err := launcher.setup()
 	if err != nil {
@@ -131,6 +133,8 @@ func (l *Launcher) run() {
 			containerID := service.Identifier
 			l.stopTailer(containerID)
 			delete(l.pendingContainers, containerID)
+		case containerId := <-l.lostSocket:
+			l.restartTailer(containerId)
 		case <-l.stop:
 			// no docker container should be tailed anymore
 			return
@@ -146,9 +150,9 @@ func (l *Launcher) startTailer(container *Container, source *config.LogSource) {
 		return
 	}
 
-	tailer := NewTailer(l.cli, containerID, source, l.pipelineProvider.NextPipelineChan())
+	tailer := NewTailer(l.cli, containerID, source, l.pipelineProvider.NextPipelineChan(), l.lostSocket)
 
-	// conpute the offset to prevent from missing or duplicating logs
+	// compute the offset to prevent from missing or duplicating logs
 	since, err := Since(l.registry, tailer.Identifier(), container.service.CreationTime)
 	if err != nil {
 		log.Warnf("Could not recover tailing from last committed offset: %v", ShortContainerID(containerID), err)
@@ -170,4 +174,27 @@ func (l *Launcher) stopTailer(containerID string) {
 		go tailer.Stop()
 		delete(l.tailers, containerID)
 	}
+}
+
+func (l *Launcher) restartTailer(containerID string) {
+	oldTailer, _ := l.tailers[containerID]
+	source := oldTailer.source
+	oldTailer.Stop()
+
+	tailer := NewTailer(l.cli, containerID, source, l.pipelineProvider.NextPipelineChan(), l.lostSocket)
+
+	// compute the offset to prevent from missing or duplicating logs
+	since, err := Since(l.registry, tailer.Identifier(), service.Before)
+	if err != nil {
+		log.Warnf("Could not recover tailing from last committed offset: %v", ShortContainerID(containerID), err)
+	}
+
+	// start the tailer
+	err = tailer.Start(since)
+	if err != nil {
+		log.Warnf("Could not start tailer: %v", containerID, err)
+	}
+
+	// keep the tailer in track to stop it later on
+	l.tailers[containerID] = tailer
 }

--- a/pkg/logs/input/docker/tailer.go
+++ b/pkg/logs/input/docker/tailer.go
@@ -144,7 +144,6 @@ func (t *Tailer) readForever() {
 				}
 				if err == io.ErrUnexpectedEOF {
 					log.Warn("Unexpected EOF: The tailer of container ", ShortContainerID(t.ContainerID), " will restart")
-					t.wait() // TODO: Implement a better retry policy
 					t.unexpectedEOF <- t.ContainerID
 				}
 				return

--- a/pkg/logs/input/docker/tailer.go
+++ b/pkg/logs/input/docker/tailer.go
@@ -31,7 +31,7 @@ const defaultSleepDuration = 1 * time.Second
 const tagsUpdatePeriod = 10 * time.Second
 
 // Tailer tails logs coming from stdout and stderr of a docker container
-// With docker api, there is no way to know if a log comes from strout or stderr
+// With docker api, there is no way to know if a log comes from stdout or stderr
 // so if we want to capture the severity, we need to tail both in two goroutines
 type Tailer struct {
 	ContainerID   string
@@ -46,10 +46,11 @@ type Tailer struct {
 	shouldStop    bool
 	stop          chan struct{}
 	done          chan struct{}
+	unexpectedEOF chan string
 }
 
 // NewTailer returns a new Tailer
-func NewTailer(cli *client.Client, containerID string, source *config.LogSource, outputChan chan message.Message) *Tailer {
+func NewTailer(cli *client.Client, containerID string, source *config.LogSource, outputChan chan message.Message, unexpectedEOF chan string) *Tailer {
 	return &Tailer{
 		ContainerID:   containerID,
 		outputChan:    outputChan,
@@ -59,6 +60,7 @@ func NewTailer(cli *client.Client, containerID string, source *config.LogSource,
 		sleepDuration: defaultSleepDuration,
 		stop:          make(chan struct{}, 1),
 		done:          make(chan struct{}, 1),
+		unexpectedEOF: unexpectedEOF,
 	}
 }
 
@@ -139,6 +141,11 @@ func (t *Tailer) readForever() {
 				if err != io.EOF {
 					t.source.Status.Error(err)
 					log.Error("Err: ", err)
+				}
+				if err == io.ErrUnexpectedEOF {
+					log.Warn("Unexpected EOF: The tailer of container ", ShortContainerID(t.ContainerID), " will restart")
+					t.wait() // TODO: Implement a better retry policy
+					t.unexpectedEOF <- t.ContainerID
 				}
 				return
 			}

--- a/releasenotes/notes/restart-tailer-when-docker-socket-is-lost-7f211a63f9158fd9.yaml
+++ b/releasenotes/notes/restart-tailer-when-docker-socket-is-lost-7f211a63f9158fd9.yaml
@@ -1,0 +1,3 @@
+fixes:
+  - |
+    Recover from errors when connection to the docker socket is lost to continue tailing containers.

